File: GMRES.py
# import packages
import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import norm

# where does the reorthogonalization step come in?
def gmres(A,b,k):
    Q,H = arnoldi(A,b,k)
    e1 = np.zeros(k)
    e1[0] = norm(b)
    y,res = np.linalg.lstsq(H[:k,:k],e1)[:2] # what is [:2] for?
    return Q[:,:k]@y

def arnoldi(A,b,k):
    # A matrix
    # b starting vector
    # k number of iterations
    # starts with n=1
    n = len(b)
    H = np.zeros((k+1,k))
    Q = np.zeros((n,k+1)) # why k+1?
    Q[:,0] = b/norm(b)
    for i in range(k):
        v = A @ Q[:,i]
        for j in range(i+1):
            q = Q[:,j]
            H[j,i] = q@v# take a bunch of inner products
            v -= H[j,i]*q # get a residual

        H[i+1,i] = norm(v)
        Q[:,i+1] = v/H[i+1,i]
    return Q,H

    # Example usage
n = 10
A = np.random.randn(n,n)
b = np.random.randn(n)
k = n

x = gmres(A,b,k)

true_x = np.linalg.solve(A,b)

print("Solution x:")
print(x)
print("True x:")
print(true_x)
print("Error:")
print(x-true_x)

File: GMRES.jl
using LinearAlgebra
using Random

# where does the reorthogonalization step come in?
function gmres(A, b, k)
    Q, H = arnoldi(A, b, k)
    e1 = zeros(k)
    e1[1] = norm(b)
    y = H[1:k, 1:k] \ e1
    return Q[:, 1:k] * y
end

function arnoldi(A, b, k)
    # A matrix
    # b starting vector
    # k number of iterations
    # starts with n=1
    n = length(b)
    H = zeros(k+1,k)
    Q = zeros(n,k+1)
    Q[:,1] = b / norm(b)
    for i in 1:k
        v = A * Q[:,i]
        for j in 1:i
            q = Q[:, j]
            H[j, i] = dot(q,v) # take a bunch of inner products
            v -= H[j,i]*q # get a residual
            # the above two lines are a reorthogonalization step... very
            # expensive.
        end
        H[i+1, i] = norm(v)
        Q[:, i+1] = v / H[i+1, i]
    end
    return Q, H
end

# Example usage
n = 10
Random.seed!(123)  # To get consistent results
A = randn(n, n)
b = randn(n)
k = n

x = gmres(A, b, k)

x_true = A \ b

println("Solution x:")
println(x)
println("True x:")
println(x_true)
println("Error:")
println((x - x_true)/norm(x_true))
